src/server/replica.cc

```diff
@@ -851,13 +851,47 @@ void Replica::RedisStreamAcksFb() {
   }
 }
 
+class AclCheckerClient : public ProtocolClient {
+ public:
+  AclCheckerClient(ServerContext server, Context* cntx)
+      : ProtocolClient(std::move(server)), cntx_(cntx) {
+    Connect();
+  }
+
+  void CheckAclRoundTrip() {
+    if (auto ec = SendCommandAndReadResponse(StrCat("REPLCONF acl-check ", "0")); ec) {
+      cntx_->Cancel();
+      LOG(INFO) << "Error in REPLCONF acl-check: " << ec.message();
+    } else if (!CheckRespIsSimpleReply("OK")) {
+      cntx_->Cancel();
+      LOG(INFO) << "Error: " << ToSV(LastResponseArgs().front().GetBuf());
+    }
+  }
+
+ private:
+  void Connect() {
+    VLOG(1) << "Connecting with acl client";
+    auto ec = ConnectAndAuth(absl::GetFlag(FLAGS_master_connect_timeout_ms) * 1ms, cntx_);
+    if (ec) {
+      LOG(INFO) << "Failed to connect with acl client " << ec.message();
+      cntx_->Cancel();
+    }
+  }
+
+  Context* cntx_;
+};
+
 void DflyShardReplica::StableSyncDflyAcksFb(Context* cntx) {
   constexpr size_t kAckRecordMaxInterval = 1024;
   std::chrono::duration ack_time_max_interval =
       1ms * absl::GetFlag(FLAGS_replication_acks_interval);
   std::string ack_cmd;
   auto next_ack_tp = std::chrono::steady_clock::now();
 
+  // We need a new client with a different socket for acl-checks
+  // because acks should not be replied to
+  AclCheckerClient acl_client(server(), cntx);
```

Comment:
This function runs for every replication connection. so you double the connections for replication now. must we have it here?

Reply:
yes so the overhead is 1 connection per replica node. However, this should not be an issue because we already doing that with `ack` and we are just sending a very small request. Yes we need it here, otherwise we can't really know if the acl's of that user changed since there is no acl validation after we reach stable sync other than `ack` which we can't use because we never reply to it. And probably we won't have more than 3 replicas per master so the average overhead is 3 extra connections

Reply:
My comment here was that we have it for every replication connection. It is not 1 extra connection per replica node but one extra connection for every shard replication connection. i.e if we have master with 32 shards we will have extra 32 connections for this replica. I was asking if we must have it here i.e if we can move it from here to another place in code that will create only one connection for the replica and not connection per replication fiber